### ISSUE TITLE
IJ-based sync takes extra time because of unexpected synchronous upload of BES metrics to remote endpoint

### ISSUE THREAD

**Issue by dkashyn-sfdc** [bazelbuild/bazel#19117]
### Description of the bug:

To not depend on BES upload timings we do have `build --bes_upload_mode=fully_async` in our `.bazelrc` file right next to `build --bes_backend=grpcs://`.

It works perfectly for CLI.

Recently we've discovered that during Sync in IJ some strange behavior is observed:
1. Bazel build reported as completed fast in 20-30s.
2. Sync can take significantly longer without any clues in Bazel output.
3. There are 3 messages, and sometimes only 2 messages `Build Completed Successfully`. 

We've added `--show_timestamps` to have a better timeline and here is what we got
<img width="1029" alt="image" src="https://github.com/bazelbuild/bazel/assets/50216138/5f044dd9-ae06-449b-a5c4-37606c79f017">

After several days of debugging here is what we've identified:
1. We were able to reproduce slowness in CLI without IJ.
2. `--output_groups=intellij-resolve-java-direct-deps,intellij-info-generic,intellij-info-java-direct-deps` caused the delay 
3. BEP file size `--build_event_binary_file=/var/folders/k3/n5f1skhj65g10072006pgtvr0000gq/T/intellij-bep-16670144-a24e-4b2b-97a2-0c5ed016e4a9` increase was identified as the alleged root cause of the delay.
4. BEP file, that is requested by IJ as part of plugin logic, can be up to 230Mb(up from 1.3mb), based on our random samples. This is quite significant, and we don't know the distribution of file sizes for our engs yet.
```
ls -l /var/folders/k3/n5f1skhj65g10072006pgtvr0000gq/T/intellij-bep-b3e5254e-b486-443e-93bf-a21aff6be98d
-rw-r--r--  1 dkashyn  staff  239361453 Jul 26 07:59 /var/folders/k3/n5f1skhj65g10072006pgtvr0000gq/T/intellij-bep-b3e5254e-b486-443e-93bf-a21aff6be98d
dkashyn@dkashyn-ltm4w4p core % ls -l /var/folders/k3/n5f1skhj65g10072006pgtvr0000gq/T/intellij-bep-b3e5254e-b486-443e-93bf-a21aff6be98d_no_output_groups 
-rw-r--r--  1 dkashyn  staff  1378801 Jul 26 08:13 /var/folders/k3/n5f1skhj65g10072006pgtvr0000gq/T/intellij-bep-b3e5254e-b486-443e-93bf-a21aff6be98d_no_output_groups
```
5. This file has 22k events in one of our scenarios`
```
jq ".id | keys | .[0]" bep_huge.json | sort | uniq -c
   1 "buildFinished"
   1 "buildMetadata"
   1 "buildMetrics"
   1 "buildToolLogs"
   2 "configuration"
10034 "namedSet"
   1 "optionsParsed"
   1 "pattern"
10663 "progress"
   1 "started"
   3 "structuredCommandLine"
 590 "targetCompleted"
 590 "targetConfigured"
   1 "unstructuredCommandLine"
   1 "workspace"
   1 "workspaceStatus"
```

This is where we stuck for quite some time and after couple more days of debugging we've identified that the following code 
```
@Option(
      name = "build_event_binary_file",
      oldName = "experimental_build_event_binary_file",
      defaultValue = "",
      implicitRequirements = {"--bes_upload_mode=wait_for_upload_complete"},
      documentationCategory = OptionDocumentationCategory.LOGGING,
      effectTags = {OptionEffectTag.AFFECTS_OUTPUTS},
      help =
          "If non-empty, write a varint delimited binary representation of representation of the"
              + " build event protocol to that file. This option implies"
              + " --bes_upload_mode=wait_for_upload_complete.")
  public String buildEventBinaryFile;
```
is the root cause of an extra delay.

There are several issues that we've identified so far:
1. BEP file size is the issue by itself.
2. We still have a delay for upload for every 'Bazel Sync` call and the file size seems quite significant for reasons out of our control.
3. Our original `build --bes_upload_mode=fully_async` was silently ignored and replaced by `--bes_upload_mode=wait_for_upload_complete` as a side effect of `build_event_binary_file` enablement by IJ.
4. Build timings reported in Bazel build output are incomplete and we found clues like `[com.google.devtools.build.lib.buildeventservice.BuildEventServiceModule.waitForBuildEventTransportsToClose] Spent 744416 milliseconds doing waiting for BES close for invocation ebc789fb-4d8f-43d6-9ebc-8ff3e11367c7` in Bazel server log alone. It was also not really easy to locate those logs clues either.

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

I don't know how to simulate slow `build --bes_backend=` but some slow server, or probably noop server that not sending back ACK events, will allow to reproduce this issue easily.

Synchronous upload can be slow for many reasons and we cannot assume that this operation is instantenious.

### Which operating system are you running Bazel on?

OSX, Ubuntu, CentOS

### What is the output of `bazel info release`?

6.1.0, 6.3.0

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.

It feels like a regression after https://github.com/bazelbuild/bazel/commit/62060babcc26f3cac3db020e423daec8755fd426

There are 3 types of BEP file output and only `binary` has the requirement for sync processing.

```
@Option(
      name = "build_event_binary_file",
      oldName = "experimental_build_event_binary_file",
      defaultValue = "",
      implicitRequirements = {"--bes_upload_mode=wait_for_upload_complete"},
      documentationCategory = OptionDocumentationCategory.LOGGING,
      effectTags = {OptionEffectTag.AFFECTS_OUTPUTS},
      help =
          "If non-empty, write a varint delimited binary representation of representation of the"
              + " build event protocol to that file. This option implies"
              + " --bes_upload_mode=wait_for_upload_complete.")
  public String buildEventBinaryFile;

...
@Option(
      name = "build_event_json_file",
      oldName = "experimental_build_event_json_file",
      defaultValue = "",
      documentationCategory = OptionDocumentationCategory.LOGGING,
      effectTags = {OptionEffectTag.AFFECTS_OUTPUTS},
      help = "If non-empty, write a JSON serialisation of the build event protocol to that file.")
  public String buildEventJsonFile;
...
@Option(
      name = "build_event_text_file",
      oldName = "experimental_build_event_text_file",
      defaultValue = "",
      documentationCategory = OptionDocumentationCategory.LOGGING,
      effectTags = {OptionEffectTag.AFFECTS_OUTPUTS},
      help =
          "If non-empty, write a textual representation of the build event protocol to that file")
  public String buildEventTextFile;
```

There are no details on why implicit sync mode was enabled in the suspect PR.

**Comment by dkashyn-sfdc** [comment]
It will be beneficial to improve Bazel output by having more clues that build is not over and Bazel stuck for too long is some service activity. 

`[com.google.devtools.build.lib.buildeventservice.BuildEventServiceModule.waitForBuildEventTransportsToClose] Spent 744416 milliseconds doing waiting for BES close for invocation ebc789fb-4d8f-43d6-9ebc-8ff3e11367c7` is so deeply hidden and not reflected in build logs and this is a bad thing by itself.

**Comment by coeuvre** [comment]
cc @meisterT

**Comment by dkashyn-sfdc** [comment]
@meisterT could you please at least share why this async-kill-pill is required for binary files?

**Comment by guw** [comment]
I can confirm the behavior with running the IJ Aspects. When on a terrible network it looks like Bazel hangs. Builds complete in few seconds but then the process hangs for a couple of minutes uploading to the server.

**Comment by coeuvre** [comment]
For `--build_event_*_file` , I think the right solution is to not apply `--bes_upload_mode=wait_for_upload_complete` when combined with `--bes_backend` so that IJ (or other IDEs) don't need to wait for the BEP upload.

For the increased BEP file size, did you mean the difference come from `not using IJ vs. using IJ`, or because of upgrading to latest Bazel version? If it is the former case, it's normal because IJ requests many outputs with aspects which will be included in the BEP. Otherwise, it should be a bug in Bazel.

**Comment by dkashyn-sfdc** [comment]
Only known to us cases of huge BEP files are when IJ (or other IDEs) aspects are present.

Yes, it seems that 
> the right solution is to not apply `--bes_upload_mode=wait_for_upload_complete`
unless this was added to make IJ happy in the first place :) Origins of this implied synchronicity is unclear from the original commit  https://github.com/bazelbuild/bazel/commit/62060babcc26f3cac3db020e423daec8755fd426 since there are no details there :(

**Comment by coeuvre** [comment]
Fixed by #19322.
